### PR TITLE
Rewrite qualifier polymorphism code so other checkers can customize.

### PR DESCRIPTION
**Pull request by smillst**
`QualifierPolymorphism` is now an interface that is implemented by `AbstractQualifierPolymorphism`.  The default implementation of `AbstractQualifierPolymorphism` is `DefaultQualifierPolymorphism`.  All of these classes are now in the `org.checkerframework.framework.type.poly`package.

The Determinism Checker also extends `AbstractQualifierPolymorphism` in [DeterminismQualifierPolymorphism](https://github.com/t-rasmud/checker-framework/blob/nondet-checker/checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java).  Supporting this customization is the reason for this rewrite.
  
Pull request #2144 should be merged first and is included in this branch.

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/framework/type/AsSuperVisitor.java [hunk 1]
```diff
@@ -603,6 +603,11 @@ public AnnotatedTypeMirror visitTypevar_Primitive(
     @Override
     public AnnotatedTypeMirror visitTypevar_Typevar(
             AnnotatedTypeVariable type, AnnotatedTypeVariable superType, Void p) {
+        // Clear the superType annotations and copy over the primary annotations before computing
```
Comment (smillst) [inline framework/src/main/java/org/checkerframework/framework/type/AsSuperVisitor.java]: The changes in this file are in PR #2142.

File: framework/src/main/java/org/checkerframework/framework/util/AnnotationMirrorMap.java [hunk 2]
```diff
@@ -24,14 +24,21 @@
 public class AnnotationMirrorMap<V> implements Map<AnnotationMirror, V> {
 
     /** The actual map to which all work is delegated. */
-    private final Map<AnnotationMirror, V> shadowMap =
-            new TreeMap<>(AnnotationUtils.annotationOrdering());
+    private final Map<AnnotationMirror, V> shadowMap;
```
Comment (smillst) [inline framework/src/main/java/org/checkerframework/framework/util/AnnotationMirrorMap.java]: (Changes in this file are under review in #2144)

File: framework/src/main/java/org/checkerframework/framework/util/AnnotationMirrorSet.java [hunk 3]
```diff
@@ -132,4 +132,16 @@ public boolean removeAll(Collection<?> c) {
     public void clear() {
         shadowSet.clear();
     }
+
```
Comment (smillst) [inline framework/src/main/java/org/checkerframework/framework/util/AnnotationMirrorSet.java]: (Changes in this file are under review in #2144)

File: framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 4]
```diff
@@ -681,8 +682,8 @@ protected void checkForDefaultQualifierInHierarchy(QualifierDefaults defs) {
      *
      * @return the QualifierPolymorphism class
      */
-    protected QualifierPolymorphism createQualifierPolymorphism() {
-        return new QualifierPolymorphism(processingEnv, this);
+    public QualifierPolymorphism createQualifierPolymorphism() {
```
[the remark this answers is not in the document]
Comment (smillst) [reply framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: We need a method that creates the new QP at least once so subclasses can change which class is used.  We could change how often that method is called, but I don't have time to do that right now.

File: checker/tests/nullness/PolyAllTest.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import org.checkerframework.dataflow.qual.Pure;
+import org.checkerframework.framework.qual.PolyAll;
+
+public class PolyAllTest<T extends Comparable<T>> {
+
+    @Pure
+    public int compare(@PolyAll T[] a1, @PolyAll T[] a2) {
```
[the remark this answers is not in the document]
Comment (smillst) [reply checker/tests/nullness/PolyAllTest.java]: Done below.

File: framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: I fixed this in https://github.com/typetools/checker-framework/pull/2143/commits/b1daabbfd9824e64d70e05a507dd567001693309

[the remark this answers is not in the document]
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: If you know of code that overrides `createQP` with a public method, you can change it back to protected.

**Pull request merged**